DSA-Campaign-Uplift-Estimation/src/main/java/com/google/sps/WebCrawler.java

```diff
@@ -0,0 +1,232 @@
+        
+        // get the significant keywords from the meta description and headers
+        HashSet<String> keywordsDescriptionHeaders = getKeywordsFromDescriptionAndHeaders(document);
+
+        /*
+         * Go through every element of keywordsDescriptionHeaders and check if the word or a close variation of it (difference of one character) is found in keywordsURLTitle.
+         * If so, add the word to sharedKeywords.
+         */
+        HashSet<String> sharedKeywords = new HashSet<String>();
+        for (String keyword : keywordsDescriptionHeaders) {
+            for (String matchingKeyword : keywordsURLTitle) {
+                if (!sharedKeywords.contains(matchingKeyword) && resembles(keyword, matchingKeyword)) {
+                    sharedKeywords.add(matchingKeyword);
+                    break;
+                }
+            }
+        }
+
+        return (((double) sharedKeywords.size()) / ((double) keywordsURLTitle.size())) + 1;
+    }
+
+    // Returns true if there is less than 1 character difference between strings str1 and str2.
+    public static boolean resembles(String str1, String str2) {
+        // check if they are the same to save time
+        if (str1.equals(str2)) {
+            return true;
+        }
+
+        // initial check to ensure that the lengths of the two strings are within 1 of each other
+        if ((str1.length() < str2.length()-1) || (str1.length() > str2.length()+1)) {
+            return false;
+        }
+
+        /*
+         * We will go from left to right through str1 and str2, comparing the characters of the two strings.
+         * posStr1 and posStr2 represent the character positions we are analyzing in str1 and str2, respectively.
+         * foundDifference will be false if, when going left to right, we haven't found a difference between the two strings yet, and true if we have found a difference.
+         */
+        boolean foundDifference = false;
+        int posStr1 = 0;
+        int posStr2 = 0;
+        while ((posStr1 < str1.length()) && (posStr2 < str2.length())) {
+            if (str1.charAt(posStr1) == str2.charAt(posStr2)) {
+                // found no difference, keep progressing through the strings
+                posStr1++;
+                posStr2++;
+            } else if (!foundDifference) {
+                // found a difference between the two strings
+                foundDifference = true;
+                
+                if ((posStr1 == str1.length()-1) || (posStr2 == str2.length()-1)) {
+                    // Case 1: We have finished traversing through either str1 or str2. This case is handled outside of the while loop, which is why we simply increment
+                    // posStr1 and posStr2 to break out of the while loop on the next iteration.
+                    posStr1++;
+                    posStr2++;
+                } else if (str1.charAt(posStr1+1) == str2.charAt(posStr2)) {
```

Comment:
What would happen in the case of resembles("abcd", "acbcd")? These have an edit distance of 1.

Reply:
That would return true. The way I designed the function was that whenever there is a one-character difference it is true. So the one character difference here is the first c in acbcd.

Reply:
Character insertions, deletions, and swaps are all one-character differences.

Reply:
Can you adding resembles("abcd", "acbcd") as a test? I don't think it works, since the code would increment posStr1 and conclude that they don't resemble each other. Whereas resembles("acbcd", "abcd") would pass.

Reply:
Yes ok I overlooked that. I rewrote the code to account for that, and I added more tests. It should run slightly faster now that I'm using substring and cutting off the while loop early (still linear time though). I also found a bug with the web crawler which I also fixed.